weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java

```diff
@@ -125,13 +126,12 @@ public void remove() {
             if (bean != null) {
                 bean.remove();
             }
+            this.removed = true;
```

Comment:
@pferraro Should this be the first line in the method? There's a bit of a race here otherwise, where the bean can be gone but isRemoved() returns false. Moving this line first there is still a race, but it's a problem only if it matters whether isRemoved() returns true before that actual removal work is done.

The only use I see of isRemoved() is in this class itself, but I think my IDE isn't finding uses of SessionObjectReference properly.

BTW, both this code and the old code don't comply with the behavior specified in the SessionObjectReference javadoc.

Reply:
> Should this be the first line in the method?

My original thinking what that, if the remove handler on the bean triggers some other invocation handled by weld, it would otherwise fail with a NoSuchEJBException.
However, I think any invocation triggered within the beans remove handler would be handled by the EJB container, rather than weld.

> There's a bit of a race here otherwise, where the bean can be gone but isRemoved() returns false. 

I think that's a valid point.  While the container is required to serialize SFSB invocations, this object resides outside of the synchronization interceptor, so technically the remove() method could be invoked by concurrent threads.

Reply:
I've updated this, such that isRemoved() will no longer return false if the bean is in the process of being, or has already been, removed.

Reply:
Thanks for the discussion, Paul.